src/utils.jl

```diff
@@ -106,3 +106,66 @@ end
 
 Distributions.support(d::FiniteDiscreteMeasure) = d.support
 Distributions.probs(d::FiniteDiscreteMeasure) = d.p
+
+"""
+    cost_matrix(
+        c,
+        μ::Union{FiniteDiscreteMeasure, DiscreteNonParametric},
+        ν::Union{FiniteDiscreteMeasure, DiscreteNonParametric}
+    )
+
+Compute cost matrix from Finite Discrete Measures `μ` and `ν` using cost function  `c`.
+
+Note that the use of functions such as `SqEuclidean()` from `Distances.jl` have
+better performance than generic functions. Thus, it's prefered to use
+`cost_matrix(SqEuclidean(), μ, ν)`, instead of `cost_matrix((x,y)->sum((x-y).^2), μ, ν)`
+or even `cost_matrix(sqeuclidean, μ, ν)`.
+
+For custom cost functions, it is necessary to guarantee that the function `c` works
+on vectors, i.e., if one wants to compute the squared Euclidean distance,
+the one must define `c(x,y) = sum((x - y).^2)`.
+
+# Example
+```julia
+μ = discretemeasure(rand(10),normalize!(rand(10),1))
+ν = discretemeasure(rand(8))
+c = TotalVariation()
+C = cost_matrix(c, μ, ν)
+```
+"""
+function cost_matrix(
+    c,
+    μ::Union{FiniteDiscreteMeasure,DiscreteNonParametric},
+    ν::Union{FiniteDiscreteMeasure,DiscreteNonParametric},
+)
+    if typeof(c) <: PreMetric && length(μ.support[1]) == 1
+        return pairwise(c, vcat(μ.support...), vcat(ν.support...))
+    elseif typeof(c) <: PreMetric && length(μ.support[1]) > 1
+        return pairwise(c, vcat(μ.support'...), vcat(ν.support'...); dims=1)
+    else
+        return pairwise(c, μ.support, ν.support)
+    end
+end
```

Comment:
I think we would want to restrict this to pairs of marginals with the same type of the support. Or at least the dimension should match in the case of arrays and scalars. So we would want a more fine-grained function signature.

In general, it would be better to avoid the type checks in the function definition since it makes it more difficult to extend and specialize the method. I think it would be better to just
- define a separate fallback for arbitrary `c` (the last branch)
- define a separate method for `c.:PreMetric`

Also one should avoid the splatting of support, it will lead to massive compile times and inference problems for larger arrays.

The same comments apply to the implementation below.

Reply:
Hmm. So the splatter is inefficient. Now, how should one efficiently construct matrix `C` using `Distances.pairwise` ? I mean, the `StatsBase.pairwise` takes the vector of vectors and returns exactly what one wants. But the `Distances.pairwise` would require a matrix version.

Reply:
As mentioned above, if you work with ColVecs or RowVecs you actually don't want to construct a matrix at all. But if you deal with an actual vector of vectors, then usually you would use e.g. `reduce(hcat, vectors_of_vectors)` to avoid splatting.

Reply:
I agree. But since we didn't require `KernelFunctions` as a dependency, I could not guarantee that `mu.support.X` would work. I'll change to your suggestion with `reduce`.